File: coding_skill_cpp/29_mock_object.cpp
/* C++ Mock Object */
#include <iostream>
using namespace std;

class ISensor {
public:
    virtual ~ISensor() = default;
    virtual float read() = 0;
};

class RealSensor : public ISensor {
public:
    float read() override {
        cout << "[Real] Hardware read" << endl;
        return 25.5f;
    }
};

class MockSensor : public ISensor {
public:
    float read() override {
        cout << "[Mock] Test data" << endl;
        return 20.0f;
    }
};

void processTemperature(ISensor& sensor) {
    float temp = sensor.read();
    cout << "[App] Temperature: " << temp << "°C" << endl;
}

int main() {
    cout << "=== C++ Mock Object ===" << endl;
    
    RealSensor real;
    MockSensor mock;
    
    processTemperature(real);
    processTemperature(mock);
    
    return 0;
}

File: coding_skill_py/29_mock_object.py
"""Python Mock Object - unittest.mock"""
from unittest.mock import Mock
from abc import ABC, abstractmethod

class ISensor(ABC):
    @abstractmethod
    def read(self) -> float:
        pass

class RealSensor(ISensor):
    def read(self) -> float:
        print("[Real] Hardware read")
        return 25.5

class MockSensor(ISensor):
    def read(self) -> float:
        print("[Mock] Test data")
        return 20.0

def process_temperature(sensor: ISensor):
    temp = sensor.read()
    print(f"[App] Temperature: {temp}°C")

def main():
    print("=== Python Mock Object ===")
    
    real = RealSensor()
    mock = MockSensor()
    
    process_temperature(real)
    process_temperature(mock)
    
    # unittest.mock 사용
    mock_sensor = Mock(spec=ISensor)
    mock_sensor.read.return_value = 22.0
    print("\n[unittest.mock]")
    temp = mock_sensor.read()
    print(f"Mocked temp: {temp}")

if __name__ == "__main__":
    main()
